src/unix/netmap-udp.c

```diff
@@ -0,0 +1,735 @@
+
+  uv__handle_init(loop, (uv_handle_t*)loop->netmap, UV_NETMAP);
+  uv__io_init(&loop->netmap->io_watcher, uv__udp_netmap_io, loop->netmap->intf->fd);
+  QUEUE_INIT(&loop->netmap->write_queue);
+  QUEUE_INIT(&loop->netmap->write_completed_queue);
+
+  return 0;
+}
+
+static uv_udp_netmap_close_cb(uv_handle_t* handle)
+{
+  handle->loop->netmap = NULL;
+  free(handle);
+}
+
+int uv_udp_netmap_close(uv_loop_t* loop) {
+  if (loop->netmap == NULL) {
+    return 0;
+  }
+
+  // XXX consider doing an ioctl to force flush the ring
+
+  loop->netmap->flags |= UV_HANDLE_CLOSING;
+  loop->netmap->close_cb = (uv_close_cb*)uv_udp_netmap_close_cb;
+  uv__io_close(loop, &loop->netmap->io_watcher);
+  uv__handle_stop(loop->netmap);
+
+  loop->netmap->next_closing = loop->closing_handles;
+  loop->closing_handles = loop->netmap;
+
+  nm_close(loop->netmap->intf);
+
+  return 0;
+}
+
+void uv_udp_netmap_set_network(uv_loop_t* loop, unsigned char* src_mac, unsigned char* dst_mac, unsigned char* src_ip) {
+  if (loop->netmap == NULL) {
+    return;
+  }
+
+  memcpy(loop->netmap->src_mac, src_mac, 6);
+  memcpy(loop->netmap->dst_mac, dst_mac, 6);
+  memcpy(loop->netmap->src_ip, src_ip, 4);
+}
+
+void uv__udp_netmap_close_handle(uv_udp_t* handle) {
+  if (handle->loop->netmap == NULL) {
+    return;
+  }
+
+  if (handle->io_watcher.fd > 0) {
+    handle->loop->netmap->sockets[handle->io_watcher.fd] = NULL;
+    handle->loop->netmap->socket_tos[handle->io_watcher.fd] = 0;
+  }
+
+  handle->io_watcher.fd = -1;
+
+  return;
+}
+
+void uv__udp_netmap_finish_close_handle(uv_udp_t* handle) {
+  uv_udp_send_t* req;
+  QUEUE* h;
+  QUEUE* q;
+  int removed;
+
+  assert(handle->io_watcher.fd == -1);
+
+  removed = 0;
+
+  if (handle->loop->netmap != NULL) {
+    if (!QUEUE_EMPTY(&handle->loop->netmap->write_queue)) {
+      h = &handle->loop->netmap->write_queue;
+      q = QUEUE_HEAD(h);
+      while (q != h) {
+        req = QUEUE_DATA(q, uv_udp_send_t, queue);
+        if (req->handle == handle) {
+          q = QUEUE_PREV(q);
+          QUEUE_REMOVE(&req->queue);
+          req->status = UV_ECANCELED;
+          QUEUE_INSERT_TAIL(&handle->loop->netmap->write_completed_queue, &req->queue);
+          removed = 1;
+        }
+
+        q = QUEUE_NEXT(q);
+      }
+
+      if (removed) {
+        uv__udp_netmap_run_completed(handle->loop);
+      }
+
+      handle->recv_cb = NULL;
+      handle->alloc_cb = NULL;
+    }
+  }
+}
+
+
+int uv__udp_netmap_bind(uv_udp_t* handle,
+                        const struct sockaddr* addr,
+                        unsigned int addrlen,
+                        unsigned int flags) {
+  if (handle->loop->netmap == NULL) {
+    return -1;
+  }
+
+  if (addrlen != sizeof(struct sockaddr_in)) {
+    return -1;
+  }
+  const struct sockaddr_in *inaddr = (const struct sockaddr_in*)addr;
+
+  if (inaddr->sin_family != AF_INET) {
+    return -1;
+  }
+
+  uint16_t port = ntohs(inaddr->sin_port);
+
+  // XXX handle 0 port (assign one)
```

Comment:
Consider an `assert` / `return -1` here if we're explicitly not supporting "give me any port."